File: tests/narcissus-test.js
#!/usr/bin/env narwhal
var narcissus = require('narcissus');

var file = require('file'),
    os = require('os'),
    json = require('json');

// Hack for v8
var BASE_DIR = '/home/andy/git/narwhal2/packages/narcissus';

function testFileContents(path) {
  var absPath = file.join(BASE_DIR, path);
  return file.open(absPath).read();
}

print("\n\n");
print("NARCISSUS");
for (var name in narcissus) {
  print(name);
}
print("\n\n");

var forLoop = "for (var i = 0; i < 10; i++) { print(i); }";

exports.testLoopParse = function() {

    // Can parse JSON Template, but can't execute it yet.
    //var code = file.open('testdata/json-template.js').read();
    print(forLoop);
    var parseTree = narcissus.parse(forLoop);

    print("Done parsing");

    if (narcissus.execute) {
      var execResult = narcissus.execute(
          parseTree, new narcissus.ExecutionContext(narcissus.GLOBAL_CODE));
      //print("execResult " + execResult);

      parseTree = narcissus.parse("function t() {return true;}");
      //print(parseTree);

      execResult = narcissus.execute(
          parseTree, new narcissus.ExecutionContext(narcissus.FUNCTION_CODE));
      //print("execResult FUNCTION_CODE " + execResult);

      parseTree = narcissus.parse(
          "function t() {print('in function t ***'); return true;}");

      execResult = narcissus.execute(
          parseTree, new narcissus.ExecutionContext(narcissus.EVAL_CODE));
      print("execResult EVAL_CODE " + execResult);
    }
}

exports.testSimpleCode = function() {
    var code = testFileContents('testdata/simple.js');

    print(code);
    var parseTree = narcissus.parse(code);
    print("Done parsing");
    //var execResult = narcissus.evaluate(code);
}

exports.testFunction = function() {
    // TODO: Make this work
    var code = testFileContents('testdata/function.js');
    //var result = narcissus.evaluate(code);
    //print("testFunction " + result);
}

exports.testParseRealCode = function() {
    var code = testFileContents('testdata/json-template.js');
    //return; // disabled for Rhino
    var parseTree = narcissus.parse(code);
    return;
    try {
      var result = narcissus.evaluate(code);
    } catch (e) {
      print('name ' + e.name);
      print('message ' + e.message);
      print('stack ' + e.stack);
    }
    print('RESULT ' + result);
}

exports.testParseItself = function() {
    //var code = testFileContents('lib/narcissus.js');
    var code = testFileContents('patched/jsdefs.js');
    var code = testFileContents('patched/jsparse.js');

    // TODO: Parse error in jsexec.js
    var code = testFileContents('patched/jsexec.js');
    try {
      var parseTree = narcissus.parse(code);
    } catch (e) {
      print('name ' + e.name);
      print('message ' + e.message);
      print('stack ' + e.stack);
    }
}

exports.testBreak = function() {
    var code = testFileContents('testdata/break.js');
    var parseTree = narcissus.parse(code);
    //print(json.stringify(parseTree, null, 2));
}

exports.testTokenizer = function() {
    var code = testFileContents('testdata/break.js');
    print(code);
    var t = new narcissus.Tokenizer(code);
    for (var i=0; i<10; i++) {
      print(t.get());
      var token = t.token();
      print(json.stringify(token));
      //print(t.token());
    }
}


if (require.main === module.id)
    require("test/runner").run(exports);

File: tests/narcissus_test.py
#!/usr/bin/python -S
"""
narcissus_test.py

Compare the output of narcissus-on-Narwhal with pynarcissus (which outputs
s-expressions on v8 testdata)
"""

__author__ = 'Andy Chu'


import commands
import glob
import os
import sys

sys.path.append('../../../../svn/pan/trunk')  # HACK
from pan.core import os_process
from pan.test import testy

PY_NARCISSUS_ROOT = '/home/andy/svn/pynarcissus-read-only'


class JsVerifier(testy.StandardVerifier):
  """Verify the JS implementation of Narcissus."""

  def __init__(self, *args, **kwargs):
    testy.StandardVerifier.__init__(self, *args, **kwargs)
    self.js_runner = os_process.Runner()
    self.py_runner = os_process.Runner(cwd=PY_NARCISSUS_ROOT)

  def _TestPath(self, path):
    return os.path.join(PY_NARCISSUS_ROOT, path)

  def _TestContents(self, path):
    return open(self._TestPath(path)).read()

  def ParseTree(self, test_file):
    js_input = self._TestContents(test_file)

    cmd = './sexp.py'
    p = self.py_runner.Pipes(cmd, 'IO')
    py_out, _ = p.communicate(js_input)

    print py_out

    js_cmd = './nw.sh $PWD/bin/narcissus %s' % self._TestPath(test_file)
    js_out = self.js_runner.Result(js_cmd).stdout
    # Hack to work around v8 spew
    js_out = js_out.split('PARSE TREE')[1].strip()
    print js_out

  def ParseTreeOfSnippet(self, snippet):
    cmd = './sexp.py'
    p = self.py_runner.Pipes(cmd, 'IO')
    py_out, _ = p.communicate(snippet)

    js_cmd = "./nw.sh $PWD/bin/narcissus -j -e %s" % commands.mkarg(snippet)
    js_out = self.js_runner.Result(js_cmd).stdout
    # Hack to work around v8 spew
    js_out = js_out.split('PARSE TREE')[1].strip()

    self.LongStringsEqual(py_out, js_out)


class SimpleTest(testy.Test):

  VERIFIERS = [JsVerifier]

  def setUpOnce(self):
    self.runner = os_process.Runner(cwd=PY_NARCISSUS_ROOT)

  def testVar(self):
    self.verify.ParseTreeOfSnippet('var x=99;')

  def testLoop(self):
    self.verify.ParseTreeOfSnippet(
        'for (var i=0; i<10; i++) { print(i); }')

  def testFunction(self):
    self.verify.ParseTreeOfSnippet(
        'function foo() { return 5; }')

  def testSmall(self):
    cmd = './jsparser.py tests/mjsunit/source/regress-1039610.js'
    print self.runner.Result(cmd).stdout

  def testSmallExample(self):
    self.verify.ParseTree('tests/mjsunit/source/regress-1039610.js')

  def testParsingAll(self):
    test_files = glob.glob(
        os.path.join(PY_NARCISSUS_ROOT, 'tests/mjsunit/source/*.js'))
    runner = os_process.Runner()
    for test_file in test_files:
      print test_file
      # TODO: output JSON, and parse it, count nodes, etc.
      js_cmd = './nw.sh $PWD/bin/narcissus %s' % test_file
      js_out = runner.Result(js_cmd).stdout
      # Hack to work around v8 spew
      js_out = js_out.split('PARSE TREE')[1].strip()
      print len(js_out)
      self.verify.IsTrue(len(js_out) > 20, js_out)

    print '%s test files' % len(test_files)


if __name__ == '__main__':
  # One-time setup
  os.system('build/makelib.py')

  testy.RunThisModule()
